### ISSUE TITLE
Swedish sorting 2006 caesura

### ISSUE THREAD

**Issue by moewew** [plk/biber#164]
Via http://tex.stackexchange.com/q/361282/35864

Apparently sorting rules in Swedish have changed in 2006. Before 2006 'V' and 'W' were to be sorted as the same character, but starting from 2006 as two different letters with 'V' sorting before 'W'.

On my Windows 10, MikTeX, `biblatex` 3.7/Biber 2.7 system. I get the pre-2006 sorting of 'V' and 'W' as the same character.

The OP and other users get another sorting where apparently 'W' sorts before all characters? (I'm waiting on confirmation about that as well as OS details - I *guess* it's Mac OS.)

```
\documentclass[swedish]{article}
\usepackage[utf8]{inputenc}
\usepackage[]{babel}
\usepackage{csquotes}
\usepackage[style=authoryear,backend=biber]{biblatex}

\usepackage{filecontents}
\begin{filecontents*}{\jobname.bib}
@book{vn,
    author = {Vn, A},
    title = {Title VN},
}
@book{vr,
    author = {Vr, A},
    title = {Title VR},
}
@book{wolf,
  author={Wolf, Ceasar},
  title={Title B},
}
\end{filecontents*}

\addbibresource{\jobname.bib} 
\DeclareNameAlias{sortname}{family-given}

\begin{document}
\nocite{*}
\printbibliography
\end{document}
```

Output:
> Vn, A (n.d.). Title VN.
Wolf, Ceasar (n.d.). Title B.
Vr, A (n.d.). Title VR.

Expected:
> Vn, A (n.d.). Title VN.
Vr, A (n.d.). Title VR.
Wolf, Ceasar (n.d.). Title B.

**Comment by perstar** [comment]
Unfortunately the situation regarding Swedish sorting rules is not as simple as that. What happened in 2006 is that a new edition of a very influential Swedish wordlist ("Svenska Akademiens ordlista") sorted V and W as separate letters for the first time. Traditionally they have seen as variants like E and É when sorting.

That doesn't automatically mean that became the only correct way to sort. The Language Planning Council of Sweden then advised that handling V and W as different letters could be used for normal words, but that *names* generally should be sorted as before. In the latest edition of "Svenska skrivregler" [= Swedish writing rules] (2008) it mentions that when sorting you can choose to sort v and w as the same or not the same depending on the wordlist; "use the system that most facilitates searching". So both ways ought to exist as variants for Swedish locales.

I realize that it's impractical to implement locales *in* Biber instead of just using them. I just want to correct the too simplistic view about what the situation is. I think many Swedes agree with that view though, and it will be more and more common that people will report software use traditional sorting as buggy. More high-profile publications differ in how they handle this. I have gone through a lot of them planning to make some overview about this.

It will be interesting to see what the coming new edition of Svenska skrivregler that will be out this spring will say!

**Comment by moewew** [comment]
Thanks for chiming in. It is good to have someone on board who knows more than one Wikipedia article on the issue.

Given that one can get the new sorting by explicitly overriding the sort locale (`sortlocale=sv-SE_reform`) and the advice on name sorting, do you think the 'reformed' sorting should be the default when using `swedish` or should Biber continue to use `sv-SE`? (Cf. Appendix A.1 of the Biber documentation).

**Comment by perstar** [comment]
I suspect that the best default might be v≠w but I would want to refer to that coming edition of Svenska skrivregler before making up my mind about that. The latest I heard it would be published "probably in April". So my suggestion is to "just" fix the bug meanwhile, and maybe mention the reform option.

**Comment by moewew** [comment]
OK, thanks.

egreg has confirmed that on a Mac with up to date packages 'W' sorts before 'I', so regardless of what we do about reformed vs non-reformed sorting, that certainly needs looking at.

**Comment by plk** [comment]
Can you confirm this last comment? I get "I" sorted before "W" with both Swedish locales on OSX.

**Comment by moewew** [comment]
Since I'm not on a Mac, I can't be sure, but I took egreg's comment (http://tex.stackexchange.com/a/361288/35864) to mean that he got 'W' before 'I'. Maybe I should ask over at TeX.SX chat if someone has a Mac and a bit of time to check.

**Comment by plk** [comment]
I'm on a Mac and it seems ok to me although I am using currently the 3.8/28 DEV versions.

**Comment by moewew** [comment]
http://chat.stackexchange.com/transcript/message/36433648#36433648

**Comment by plk** [comment]
Can you get them to try 3.8/2.8 dev versions on osx?

**Comment by eg9** [comment]
I did a bit more experimenting and tried

    \documentclass[swedish]{article}
    \usepackage[utf8]{inputenc}
    \usepackage[]{babel}
    \usepackage{csquotes}
    \usepackage[style=authoryear,backend=biber]{biblatex}
    
    \usepackage{filecontents}
    \begin{filecontents*}{\jobname.bib}
    @book{aa,
        author = {Aachen, Georgina},
        title = {Title AA},
    }
    @book{vn,
        author = {Vn, A},
        title = {Title VN},
    }
    @book{vr,
        author = {Vr, A},
        title = {Title VR},
    }
    @book{wolf,
      author={Wolf, Ceasar},
      title={Title B},
    }
    @book{ringa,
        author = {Ånten, Georgina},
        title = {Title Å},
    }
    @book{ua,
        author = {Änten, Georgina},
        title = {Title Ä},
    }
    @book{uo,
        author = {Önten, Georgina},
        title = {Title Ö},
    }
    \end{filecontents*}
    
    \addbibresource{\jobname.bib} 
    \DeclareNameAlias{sortname}{family-given}
    
    \begin{document}
    \nocite{*}
    \printbibliography
    \end{document}

The result is

<img width="325" alt="screen shot 2017-04-01 at 16 06 40" src="https://cloud.githubusercontent.com/assets/302839/24579406/395149aa-16f5-11e7-9b82-5e56802f6be1.png">

Here's the `.blg`

    [0] Config.pm:354> INFO - This is Biber 2.7
    [0] Config.pm:357> INFO - Logfile is 'moewe.blg'
    [54] biber-darwin:303> INFO - === Sat Apr  1, 2017, 16:02:36
    [80] Biber.pm:359> INFO - Reading 'moewe.bcf'
    [194] Biber.pm:832> INFO - Using all citekeys in bib section 0
    [211] Biber.pm:3670> INFO - Processing section 0
    [231] Biber.pm:3840> INFO - Looking for bibtex format file 'moewe.bib' for section 0
    [243] bibtex.pm:1435> INFO - Decoding LaTeX character macros into UTF-8
    [247] bibtex.pm:1292> INFO - Found BibTeX data source 'moewe.bib'
    [289] UCollate.pm:68> INFO - Overriding locale 'sv-SE' defaults 'variable = shifted' with 'variable = non-ignorable'
    [289] UCollate.pm:68> INFO - Overriding locale 'sv-SE' defaults 'normalization = NFD' with 'normalization = prenormalized'
    [289] Biber.pm:3499> INFO - Sorting list 'nyt/global/' of type 'entry' with scheme 'nyt' and locale 'sv-SE'
    [299] bbl.pm:608> INFO - Writing 'moewe.bbl' with encoding 'UTF-8'
    [301] bbl.pm:712> INFO - Output to moewe.bbl

**Comment by moewew** [comment]
Note sure if that helps.

Christian Hupfer (who gets incorrect sorting on Linux) reports that he gets `locale_version` 1.14 in the Biber `--trace`, my trace reads 1.19 and sorts correctly. (http://chat.stackexchange.com/transcript/message/36433978#36433978)

**Comment by plk** [comment]
Are they using pure perl biber? This could just be a module version issue.

**Comment by moewew** [comment]
Mhhh, Unicode::Collate seems to be at version 1.19 currently. http://search.cpan.org/~sadahiro/Unicode-Collate-1.19/Collate.pm Maybe 1.14 that Christian is reporting is not good enough.

**Comment by moewew** [comment]
Sorry, didn't read your last comment. I'm not sure, I expect they use what ships with TeX live. What version is that? How could one tell?

**Comment by u-fischer** [comment]
I get "correct" sorting (W between the both V) with miktex but wrong (W at the start) with texlive 2016. Both on win10. I will try with a biber 2.8. asap.

**Comment by u-fischer** [comment]
Imho the reason of the problem is as moewew suspected the locale version. The files in the biber cache in  lib/Unicode/Collate/Locale have version 1.14 in texlive 16, but 1.19 with biber-dev (2.8) and miktex. As a test I copied the Locale folder from miktex to the texlive cache and then the texlive sorting was correct.

**Comment by plk** [comment]
This is almost certainly an issue with the Unicode::Collate version. They regularly update tailorings and I always use the latest version with a Biber release.

**Comment by moewew** [comment]
Question is why TeX live's Biber (2.7) seems to consistently ship only version 1.14 of Unicode::Collate, while MikTeX uses 1.19.

**Comment by plk** [comment]
It could be due to when the pulls of the code were done. I am not sure but perhaps MikTeX pulled later and there was a correction.

**Comment by moewew** [comment]
Mhhh, interestingly, the 2.7 release requires 1.19, but you only stepped up the requirement from 1.18 on 2016-12-05 (https://github.com/plk/biber/commit/4bd806a6ff526d7a32ec991c62044f4b14ab0297). Before that it went up from 1.14 to 1.18 on 2016-11-11 (https://github.com/plk/biber/commit/3d62817080598d68514ed2691b1f616fcda2abe3). That would mean that TeX live people build a version considerably earlier than official release (almost a month). Odd...

**Comment by plk** [comment]
I've had that problem before - they don't wait until I tell them all the binaries are ready and build as soon as they see a new version.

**Comment by moewew** [comment]
Mhhh, but wouldn't that mean that they effectively build a beta version that identifies as such? You only set the `our $BETA_VERSION = 0;` on 2016-12-05 (https://github.com/plk/biber/commit/4bd806a6ff526d7a32ec991c62044f4b14ab0297).

Maybe there is an official way to tell them which source to use? (I mean, how would they know when to build a new release if you haven't released it officially yet. The time difference here is one month...)

**Comment by plk** [comment]
Sometimes what happens is that I release a version on SF and upload the binaries I build myself. Then I ask the other binary builders to build and upload. Then I announce to the CTAN maintainers to pull. A couple of times, there has been quite a time for the other builders to upload and it seems that the CTAN maintainers get wind of this and pull before I tell them. I suspect what happened is that a new U::C version came out and I updated again but they had already pulled.

**Comment by moewew** [comment]
Mhhh, the sources on CTAN right now want version 1.18 of Unicode::Collate, that is not actually what the release version on Github wants, but more than 1.14.

Anyhow, I suspect this is a weird one-off. We should probably check back with people once 2.8 is released and on CTAN/TeX live/MikTeX, but until then, there is not a lot we can do.

**Comment by plk** [comment]
There was an issue with the OSX binary due to `Unicode::Collate` 1.19 switching install paths on OSX. This should be resolved in the 2.8 dev pre-compiled version.

**Comment by moewew** [comment]
Just because it came up again on TeX.SX. TeX live 2017's Biber 2.7 still does not sort correctly (see e.g. https://tex.stackexchange.com/questions/164758/wrong-alphabetic-order-for-norwegian-bibliography-with-biblatex?noredirect=1&lq=1#comment985383_164763) I assume this is because Unicode::Collate is still in the wrong version, but have not yet gotten confirmation.

**Comment by perstar** [comment]
It's not coming up "again". It has been a constant problem for many texlive users all this time, as far as I know. According to comments in #168 that will not be changed.

**Comment by plk** [comment]
Yes, this was a bug with 2.7 - the correct version of Unicode::Collate was included but on OSX, the U::C maintainers switched the install path (again) and since there is always a default version with perl anyway, OSX picked up the earlier default version instead. This is fixed with 2.8 which I hope to release soonish.

**Comment by perstar** [comment]
I never thought #168 was only about OSX (and that's why I confirmed how it was not working for me as well there). But with the bug closed and with the impression that the latest version of RHEL/Centos is seen as too old to bother about I have had to stop evangelizing about Biber at my department for the time being. (I'm almost sure I will eventually be able to start recommending it again, though.)

**Comment by moewew** [comment]
Mhhh, AFAIK the problem was also present with Linux. Above Ulrike mentions that the she could reproduce the wrong sorting even on Windows with TeX live. So the problem does not seem to be the OS, but rather that TeX live shipped an outdated version. The question is why? If they pulled the latest code for Biber 2.7 without modification Unicode::Collate should be newer than 1.14.

**Comment by moewew** [comment]
Mhhh, AFAIK the problem was also present with Linux. Above Ulrike mentions that the she could reproduce the wrong sorting even on Windows with TeX live. So the problem does not seem to be the OS, but rather that TeX live shipped an outdated version. The question is why. If they pulled the latest code for Biber 2.7 without modification Unicode::Collate should be newer than 1.14.

It would be really annoying if we thought the bug was fixed because it didn't appear in the sourceforge release, when there is something in TeX live's Biber that makes it use older versions of Unicode::Collate and thus causes the bug to reappear.

**Comment by u-fischer** [comment]
I still get the wrong sorting with texlive on windows. Collate.pm has version 1.19, but the locale files, e.g. nn.pm, have the version 1.14 but in miktex their version is also 1.19. With a newer biber that I have around (2.7 beta) it works fine.

**Comment by sieversMartin** [comment]
Having the same `biber` and `biblatex` versions installed on MiKTeX and TeXLive and getting different result when using `sortlocale` is really a miserable situation IMHO.

When is v2.8 expected to be released? Instead of waiting, could you maybe clone the "real" 2.7 to 2.7a and tell the CTAN team to grab that one? As they obviously took the wrong files, another way would be to ask them to grab the most recent 2.7 and substitute their 2.7. Wdyt?

**Comment by plk** [comment]
I’d like to release the new versions very soon - this weekend ideally - @moewew - any objection?

**Comment by moewew** [comment]
There are a few unresolved questions in https://github.com/plk/biblatex/issues/627 and https://github.com/plk/biblatex/issues/520 that only need a few quick decisions. 

The other open issues either need more investigation or can be deferred easily.

So we could release soon. If we get along with the other issues well, we might be able to push out another version not too long after that, but that is not a bad thing in my opinion. If we release now we have roughly half a year until the next TeX live release, so enough time to fix things if Biber packaging goes wrong again or anything else crops up. 

But I don't think we can release right away. The name change of `\DeclareSortingScheme`  affects many styles (on my machine alone `archaeologie`, `biblatex-abnt`, `biblatex-anonymous`, `biblatex-apa` (not that big of a problem), `biblatex-archaeology`, `biblatex-arthistory-bonn`, `biblatex-caspervector`, `biblatex-chicago`, `biblatex-claves`, `biblatex-fiwi`, `biblatex-gb7714-2015`, `biblatex-gost`, `biblatex-manuscripts-philology`, `biblatex-oxref`, `biblatex-publist`, `biblatex-realauthor`, `biblatex-sbl`, `biblatex-swiss-legal`, `droit-fr`, `oscola`). As far as I can see there is no backwards compatibility in place, just a deprecation warning. That would be a major source of trouble.

There are other breaking changes that also require a bit of advance warning (we had the `origlanguage` change).

I think we should announce plans to publish a new version, let's say, two weeks before the update is actually sent to CTAN along with the expected changes, especially the breaking changes, on comp.text.tex and other channels, so style developers can take measures to address problems that might arise.

**Comment by plk** [comment]
The `*Scheme` changes will be backwards compatible with a warning - I just pushed that change. The `origlanguage` change I think is not so bad - we already spoke to the main style authors and the main two said it didn't matter much to them. But, I agree, we can make an announcement to CTAN very soon with a pointer to the changes file which I'll update later today.

**Comment by moewew** [comment]
OK, then the breaking changes are not that dangerous after all. A short notice period is still called for, but it probably does not need to be that long.

**Comment by ndlecic** [comment]
Sorry for jumping into this thread, but since you discuss 2.8 release, please make sure that you include Lingua::Translit 0.28; I submitted one important correction to the Russian transliteration used by biber.

**Comment by plk** [comment]
Thanks for letting me know - I will update the dependency in biber right now.

**Comment by plk** [comment]
I will close this as we are going to release 2.8 soon.
